File: apps/api/app/schemas/task.py
"""Pydantic Schemas: Task and TaskDependency"""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict
from app.models.enums import TaskStatus, TaskPriority, DependencyType


class TaskBase(BaseModel):
    name: str
    description: Optional[str] = None
    status: str = TaskStatus.PENDING.value
    priority: str = TaskPriority.MEDIUM.value
    planned_start: Optional[datetime] = None
    planned_end: Optional[datetime] = None
    actual_start: Optional[datetime] = None
    actual_end: Optional[datetime] = None


class TaskCreate(TaskBase):
    pass


class TaskUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    priority: Optional[str] = None
    planned_start: Optional[datetime] = None
    planned_end: Optional[datetime] = None
    actual_start: Optional[datetime] = None
    actual_end: Optional[datetime] = None


class TaskResponse(TaskBase):
    id: str
    event_id: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TaskDependencyCreate(BaseModel):
    predecessor_task_id: str
    successor_task_id: str
    dependency_type: str = DependencyType.FINISH_TO_START.value


class TaskDependencyResponse(BaseModel):
    id: str
    event_id: str
    predecessor_task_id: str
    successor_task_id: str
    dependency_type: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: packages/contracts/task.ts
import { TaskStatus, DependencyType } from "./enums";

export interface Task {
  id: string;
  eventId: string;
  title: string;
  description: string;
  status: TaskStatus;
  plannedStart: string;
  plannedEnd: string;
  actualStart?: string;
  actualEnd?: string;
  durationMinutes: number;
  slackMinutes: number;
  isCriticalPath: boolean;
}

export interface TaskDependency {
  id: string;
  eventId: string;
  predecessorId: string;
  successorId: string;
  dependencyType: DependencyType;
  lagMinutes: number;
}
